python/ray/ml/train/gbdt_trainer.py

```diff
@@ -0,0 +1,189 @@
+        gpus_per_worker = scaling_config_dataclass.num_gpus_per_worker
+
+        ray_params = self._ray_params_cls(
+            num_actors=int(num_workers),
+            cpus_per_actor=int(cpus_per_worker),
+            gpus_per_actor=int(gpus_per_worker),
+            resources_per_actor=resources_per_worker,
+            **self._default_ray_params,
+        )
+
+        return ray_params
+
+    def _load_model_from_checkpoint(self):
+        raise NotImplementedError
+
+    def _train(self, **kwargs):
+        raise NotImplementedError
+
+    @property
+    def _ray_params(self) -> xgboost_ray.RayParams:
+        return self._convert_scaling_config_to_ray_params(self.scaling_config)
+
+    def preprocess_datasets(self) -> None:
+        super().preprocess_datasets()
+
+        # XGBoost/LightGBM-Ray requires each dataset to have at least as many
+        # blocks as there are workers.
+        # TODO: Move this logic to the respective libraries
+        for dataset_key, dataset in self.datasets.items():
+            if dataset.num_blocks() < self._ray_params.num_actors:
+                warnings.warn(
+                    f"Dataset '{dataset_key}' has {dataset.num_blocks()} blocks, "
+                    f"which is less than the `num_workers` "
+                    f"{self._ray_params.num_actors}. "
+                    f"This dataset will be automatically repartitioned to "
+                    f"{self._ray_params.num_actors} blocks."
+                )
+                self.datasets[dataset_key] = dataset.repartition(
+                    self._ray_params.num_actors
+                )
+
+    def training_loop(self) -> None:
+        config = self.train_kwargs.copy()
+
+        dmatrices = self._get_dmatrices(
+            dmatrix_params=self.dmatrix_params,
+        )
+        train_dmatrix = dmatrices[TRAIN_DATASET_KEY]
+        evals_result = {}
+
+        init_model = None
+        if self.resume_from_checkpoint:
+            init_model = self._load_model_from_checkpoint()
+
+        config.setdefault("verbose_eval", False)
+        config.setdefault("callbacks", [])
+        config["callbacks"] += [
+            self._tune_callback_cls(filename=MODEL_KEY, frequency=1)
+        ]
+        config[self._init_model_arg_name] = init_model
+
+        self._train(
+            params=self.params,
+            dtrain=train_dmatrix,
+            evals_result=evals_result,
+            evals=[(dmatrix, k) for k, dmatrix in dmatrices.items()],
+            ray_params=self._ray_params,
+            **config,
+        )
+
+    def as_trainable(self) -> Type[Trainable]:
+        trainable_cls = super().as_trainable()
+        scaling_config = self.scaling_config
+
+        class GBDTTrainable(trainable_cls):
+            @classmethod
+            def default_resource_request(cls, config):
+                updated_scaling_config = config.get("scaling_config", scaling_config)
+                return self._convert_scaling_config_to_ray_params(
```

Comment:
We can do this in a follow up, but can we avoid serializing `self` inside the Trainable? I think we can just make `_convert_scaling_config_to_ray_params` into a pure function, create a partial, and then call the partial directly instead of directly calling `self`.

Reply:
sure